rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+
+| action       | initial state | change                           |
+| ------------ | ------------- | -------------------------------- |
+| role removal | _admin_       | _admin_ tries to remove own role |
+
+The error message is displayed: _failed to delete role that still in use by a user. Check system server logs for more details_.
+
+There is one more case when a user can lock himself out:
+
+| initial state | change                                                                                    |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |
+
+However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.
+
+## Terminology
+
+- _admin_: local user with user/role editing capabilities
+- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)
+
+## Why
+
+When a user locks themselves out, they are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.
+
+On-prem customers could work around this "locked out" state by running `tctl` locally on the auth server to create a new admin user.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can have full access.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least _admin_
+
+This will ensure that:
+
+- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
+- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
+- Locking will be impossible on the SSO side since there will be always _admin_ on our side.
+
+Now we need to define what _admin_ means in our system. W can take 2 approaches:
+
+1. User with an immutable `editor` role is _admin_.
```

Comment:
Have we considered having a magical "admin"-like user that can never be deleted, akin to a Linux root?

Given that it's likely a common idea, I think it's worth having a small section in the design comparing our solution to it and saying why we didn't go this way.

Reply:
Yeah, I have considered it. The problem is that customers would need to add this root user to each cluster.

Reply:
That sounds doable, though.

I wish we registered a range of user names for system use. What's done is done, but maybe we could take advantage of your RFD and introduce that rule for 11+ clusters.

Reply:
@xinding33 What do you think?